### PR TITLE
1412 task3 secure web api with authenticationand authorization

### PR DESCRIPTION
**Pull request by MuhammadFaizanGulzar**

### CODE REVIEW

File: CRUD.Domain/Models/AuthResponse.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CRUD.Domain.Models
+{
+    public class AuthResponse
+    {
+        public string? Token { get; set; }
```
Comment (taimoor0nline) [inline CRUD.Domain/Models/AuthResponse.cs]: why token and message are nullable?

File: CRUD.Domain/Class1.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿namespace CRUD.Domain
+{
```
Comment (taimoor0nline) [inline CRUD.Domain/Class1.cs]: remove this class if you are not using it

File: CRUD.Application/Class1.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace CRUD.Application
+{
+    public class Class1
+    {
```
Comment (taimoor0nline) [inline CRUD.Application/Class1.cs]: remove this class if you are not using it

File: CRUD.Domain/Models/Authentication/Login.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
```
Comment (taimoor0nline) [inline CRUD.Domain/Models/Authentication/Login.cs]: Please remove it if you are not using it

File: CRUD.Domain/Models/Authentication/SignUp.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CRUD.Domain.Models.Authentication
+{
+    public class SignUp
```
Comment (taimoor0nline) [inline CRUD.Domain/Models/Authentication/SignUp.cs]: Please remove it if you are not using it

File: CRUD.Domain/Models/LoginRequest.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CRUD.Domain.Models
+{
+    public class LoginRequest
+    {
+        public string? username { get; set; }
```
Comment (taimoor0nline) [inline CRUD.Domain/Models/LoginRequest.cs]: why the user and password are nullable?

File: CRUD.Domain/Models/RegisterUser.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CRUD.Domain.Models
+{
+    public class RegisterUser
+    {
+        [Required]
+        public string? userName { get; set; }
+
```
Comment (taimoor0nline) [inline CRUD.Domain/Models/RegisterUser.cs:14]: you have marked it as nullable and marking it as required as well ?

File: CRUD.Domain/Models/Response.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CRUD.Domain.Models
+{
+    public class Response
+    {
+        public string? Status { get; set; }
```
Comment (taimoor0nline) [inline CRUD.Domain/Models/Response.cs]: response should not be nullable there should be status and message

File: CRUD.Domain/Models/Role.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CRUD.Domain.Models
+{
+    public class Role : IdentityRole<int>
+    {
+        public int? Id { get; set; }
```
Comment (taimoor0nline) [inline CRUD.Domain/Models/Role.cs]: Please fix these nullable, should ID be nullable?

File: CRUD.Domain/Models/User.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+
+namespace CRUD.Domain.Models
+{
+    public class User : IdentityUser<int>
+    {
+        [Required]
+
+        public string? Email { get; set; }
```
Comment (taimoor0nline) [inline CRUD.Domain/Models/User.cs]: fix nullable email

**Review by taimoor0nline — CHANGES_REQUESTED** [review]
Please review and fix code

File: WebApiCRUD/Controllers/TodoItemController.cs [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+
+            if (todo == null)
+            {
+                return NotFound(); // Return status 404 if the todo item doesn't exist
+            }
+
+            // Get the user's roles from the claims
+            var userRoles = _httpContextAccessor.HttpContext.User.FindAll(ClaimTypes.Role).Select(c => c.Value).ToList();
+
+            if (userRoles.Contains("Admin") || (userRoles.Contains("User") && todo.UserId == userId))
```
Comment (Gulzaib83) [inline WebApiCRUD/Controllers/TodoItemController.cs]: You should implement a middleware to implement the logic "One user should only perform CRUD over its own ToDos". If a particular Todo is not under ownership of a user then it Middleware should intercept this call and it should not reach the Controller"

File: WebApiCRUD/appsettings.json [hunk 12]
```diff
@@ -0,0 +1,16 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Information",
+      "Microsoft.AspNetCore": "Warning"
+    }
+  },
+  "AllowedHosts": "*",
+  "ConnectionStrings": {
+    "DefaultConnection": "Server=.;Database=CRUD;TrustServerCertificate=True;Integrated Security=true; MultipleActiveResultSets=true"
```
Comment (Gulzaib83) [inline WebApiCRUD/appsettings.json:10]: See if you can implement "Options" design pattern for DB ConnectionString and JWT key.

**Pull request closed without merging**